instrumentation/vertx-web-3.0/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/vertx/HttpRequestInstrumentation.java

```diff
@@ -0,0 +1,261 @@
+import java.util.HashMap;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.hypertrace.agent.core.config.HypertraceConfig;
+import org.hypertrace.agent.core.instrumentation.HypertraceSemanticAttributes;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedBuffersFactory;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedCharArrayWriter;
+import org.hypertrace.agent.core.instrumentation.utils.ContentTypeUtils;
+
+public class HttpRequestInstrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<ClassLoader> classLoaderOptimization() {
+    return hasClassesNamed("io.vertx.core.http.HttpClientRequest");
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return implementsInterface(named("io.vertx.core.http.HttpClientRequest"));
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    Map<ElementMatcher<? super MethodDescription>, String> transformers = new HashMap<>();
+
+    transformers.put(
+        isMethod().and(named("write").and(takesArgument(0, is(String.class)))),
+        HttpRequestInstrumentation.class.getName() + "$WriteRequestAdvice_string");
+    transformers.put(
+        isMethod()
+            .and(named("write").and(takesArguments(1)).and(takesArgument(0, is(Buffer.class)))),
+        HttpRequestInstrumentation.class.getName() + "$WriteRequestAdvice_buffer");
+
+    transformers.put(
+        isMethod().and(named("end").and(takesArguments(0))),
+        HttpRequestInstrumentation.class.getName() + "$EndRequestAdvice");
+    transformers.put(
+        isMethod().and(named("end").and(takesArgument(0, is(String.class)))),
+        HttpRequestInstrumentation.class.getName() + "$EndRequestAdvice_string");
+    transformers.put(
+        isMethod().and(named("end").and(takesArguments(1)).and(takesArgument(0, is(Buffer.class)))),
+        HttpRequestInstrumentation.class.getName() + "$EndRequestAdvice_buffer");
+
+    return transformers;
+  }
+
+  public static class EndRequestAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(@Advice.This HttpClientRequest request) {
+
+      int callDepth = CallDepthThreadLocalMap.incrementCallDepth(HttpClientRequest.class);
+      if (callDepth > 0) {
+        return;
+      }
+
+      Contexts contexts =
+          InstrumentationContext.get(HttpClientRequest.class, Contexts.class).get(request);
+      if (contexts == null) {
+        return;
+      }
+      Span span = Span.fromContext(contexts.context);
+
+      BoundedCharArrayWriter buffer =
+          InstrumentationContext.get(MultiMap.class, BoundedCharArrayWriter.class)
+              .get(request.headers());
+      if (buffer != null) {
```

Comment:
Dont we want to check if 
```
(HypertraceConfig.get().getDataCapture().getHttpBody().getRequest().getValue()
          && ContentTypeUtils.shouldCapture(contentType))

Reply:
we don't have to check it here bc the buffer is null if the body shouldn't be captured.